lib/couchbase/protostellar/error_handling.rb

```diff
@@ -0,0 +1,144 @@
+  module Protostellar
+    module ErrorHandling
+      TYPE_URL_PRECONDITION_FAILURE = "type.googleapis.com/google.rpc.PreconditionFailure"
+      TYPE_URL_RESOURCE_INFO = "type.googleapis.com/google.rpc.ResourceInfo"
+      TYPE_URL_ERROR_INFO = "type.googleapis.com/google.rpc.ErrorInfo"
+      TYPE_URL_BAD_REQUEST = "type.googleapis.com/google.rpc.BadRequest"
+
+      def self.handle_grpc_error(grpc_error, request)
+        request.context = {} if request.context.nil?
+        detail_block = {}
+
+        rpc_status = grpc_error.to_rpc_status
+        request.context[:grpc_error] = {type: grpc_error.class}
+        request.context[:grpc_error][:code] = rpc_status.code unless rpc_status.nil?
+
+        # Decode the detail block
+        rpc_status&.details&.each do |d|
+          type_url = d.type_url
+
+          if type_url == TYPE_URL_PRECONDITION_FAILURE
+            precondition_failure = Google::Rpc::PreconditionFailure.decode(d.value)
+            detail_block[:precondition_failure] = precondition_failure
+            context[:precondition_violations] = precondition_failure.violations.map do |v|
+              {
+                violation_type: v.type,
+                violation_subject: v.subject,
+                violation_description: v.description,
+              }
+            end
+          end
+
+          if type_url == TYPE_URL_RESOURCE_INFO
+            resource_info = Google::Rpc::ResourceInfo.decode(d.value)
+            detail_block[:resource_info] = resource_info
+            request.context[:resource_type] = resource_info.resource_type
+            request.context[:resource_name] = resource_info.resource_name
+          end
+
+          if type_url == TYPE_URL_ERROR_INFO
+            error_info = Google::Rpc::ErrorInfo.decode(d.value)
+            detail_block[:error_info] = error_info
+            request.context[:error_reason] = error_info.reason
+            request.context[:error_domain] = error_info.domain
+            request.context[:error_metadata] = error_info.metadata.to_h
+          end
+
+          if type_url == TYPE_URL_BAD_REQUEST
+            bad_request = Google::Rpc::BadRequest.decode(d.value)
+            detail_block[:bad_request] = bad_request
+            request.context[:field_violations] = bad_request.field_violations.map do |v|
+              {
+                field: v.field,
+                description: v.description,
+              }
+            end
+          end
+        end
+
+        case request.service
+        when :kv
+          handle_kv_error(grpc_error, detail_block, request)
+        when :query
+          handle_query_error(grpc_error, detail_block, request)
+        else
+          handle_generic_error(grpc_error, detail_block, request)
+        end
+      end
+
+      def self.handle_generic_error(grpc_error, detail_block, request)
+        case grpc_error
+        when GRPC::DeadlineExceeded
+          RequestBehaviour.fail(Couchbase::Error::Timeout.new(nil, request.context))
```

Comment:
Locally I modified the constructor of the Classic SDK's errors to be able to take the context as an optional argument - so it should not affect the behaviour of the Classic SDK (I think). Should I push that change to the classic SDK and keep this as it is here, or use `instance_variable_set` instead?